compiler/simplCore/SimplMonad.hs

```diff
@@ -249,4 +255,69 @@ freeTick :: Tick -> SimplM ()
 -- used to decide when nothing further has happened
 freeTick t
    = SM (\_st_env us sc -> let sc' = doFreeSimplTick t sc
-                           in sc' `seq` return ((), us, sc'))
+                           in sc' `seq` return ((), [], us, sc'))
+
+{-
+************************************************************************
+*                                                                      *
+Weights
+*                                                                      *
+************************************************************************
+-}
+
+{- Note [Scaling lets]
+The simplifier creates let bindings under certain circumstances which
+are then inserted later. These are returned in `SimplFloats`.
+
+However, we have to be somewhat careful here when it comes to linearity
+as if we create a floating binding x in the scrutinee position.
+
+case_w (let x[1] = "Foo" in Qux x) of
+  Qux _ -> "Bar"
+
+then the let will end up outside the `case` if we perform KnownBranch or
+the case of case optimisation.
+
+let x[1] = "Foo"
+in "Bar"
+
+So we get a linearity failure as the one usage of x is eliminated.
+However, because the ambient context is an Omega context, we know that
+we will use the scrutinee Omega times and hence all bindings inside it
+Omega times as well. The failure was that we created a [1] binding
+whilst inside this context and it then escaped without being scaled.
+
+We also have to be careful as if we have a [1] case
+
+case_1 (let x[1] = "Foo" in Qux x) of
+  Qux x -> x
+
+then the binding maintains the correct linearity once it is floated rom
+the case and KnownBranch is performed.
+
+let x[1] = "Foo"
+in x
+
+The difficulty mainly comes from that we only discover this context
+at a later point once we have rebuilt the contituation. So, whilst rebuilding
+a continuation we keep track of how many case-of-case like opportunities
+take place and hence how much we have to scale lets floated from the scrutinee.
+This is achieved by adding a Writer like effect to the SimplM data type.
+It seems to work in practice, at least for T12944 which originally highlighted
+this problem.
+
+Why do we do this scaling afterwards rather than when the binding is
+created? It is possible the binding comes from a point deep inside the
+expression. It wasn't clear to me that we know enough about the context
+at the point we make the binding due to the SimplCont type. It might
+be thread this information through to get it right at definition site.
+For now, I leave warnings and this message to my future self.
```

Comment:
I seems to be that this strategy will have troubles with cases like this:

```haskell

f :: T :q-> T
g :: T :p-> T

case_p
  let x[1] = "X" in Foo $ "x" <>
    case_q
      let y[1] = "Y" in Bar y
    of Bar y' -> f y'
of 
Foo x' -> g x'
```

Where, after floating lets, `x` should have multiplicity `p`, while `y` should have multiplicity `pq`. But you seem to give them the same multiplicity. Am I correct?

Reply:
I suspect this will work out ok, but I don't have a test to prove it and am finding it hard to reason about why.